compiler/src/dotty/tools/dotc/core/SymDenotations.scala

```diff
@@ -1648,9 +1648,7 @@ object SymDenotations {
             if (subsym eq symbol) tp
             else subsym.denot match {
               case clsd: ClassDenotation =>
-                val tparams = clsd.typeParams
-                if (tparams.hasSameLengthAs(args)) baseTypeOf(tycon).subst(tparams, args)
-                else NoType
+                baseTypeOf(tycon).substTypeParams(tycon, args)
```

Comment:
I think the real fault is the condition. It was meant to check that we have an application of a class instead of a type lambda but failed at that. So I'd fix that first. One candidate is the condition of `substTypeParams`, that the `typeParams` of `subsym` are symbols. If that's the case, proceed as before, if not do a `baseTypeOf(superType)`, again as before.

Reply:
Makes sense. PR updated.